[Feature] Enable prometheus metrics for local queues (#2516)
kubernetes-sigs/kueue: keps/1833-metrics-for-local-queue/README.md

1. If `EnableLocalQueueResources` is true, and `LocalQueueMetricOptions` is **not** provided - metrics will be exposed for all local queues.
1. If `EnableLocalQueueResources` is true and `LocalQueueMetricOptions`  is provided - metrics will be collected for local queues that match specified label selectors.

### List of metrics for Local Queues:

In the first iteration, following are the list of metrics that would contain information on Local Queue statuses:

| Metrics Name                   | Prometheus Type | Description                                                 | 
|--------------------------------|-----------------|-------------------------------------------------------------|
| local_queue_pending_workloads  | Gauge           | The number of pending workloads                             |
| local_queue_reserved_workloads | Counter         | Total number of workloads in the LocalQueue reserving quota |
| local_queue_admitted_workloads | Counter         | Total number of admitted workloads                          |
| local_queue_resource_usage     | Gauge           | Total quantity of used quota per resource for a Local Queue |

> 

Reviewer: Could we also add metrics such as `local_queue_evicted_workloads`, `local_queue_admission_wait_time_seconds`,  `local_queue_reserved_wait_time_seconds` similarly to https://kueue.sigs.k8s.io/docs/reference/metrics/#clusterqueue-status

Author: @PBundyra I was hoping to add this in the next iteration, to make the initial set minimal and just introduce this feature. But probably, worth adding this to the KEP and just separating it out during implementation to make it easier for reviews.

Author: Addressed in https://github.com/kubernetes-sigs/kueue/pull/2516/commits/e86c944647715169fc388477e72c75051694d480

Reviewer: Thank you, can we also add `local_queue_wait_time_seconds`?

Author: Could that not be covered by admission and reserved wait time parameters?

Reviewer: It would, but I imagine it would be hard to sum up those times to get aggregated results

Author: Updated.

  - [Test Plan](#test-plan)
    - [Unit Tests](#unit-tests)
    - [Integration tests](#integration-tests)
  - [Graduation Criteria](#graduation-criteria)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
<!-- /toc -->

## Summary

The enhancement aims to introduce the exposure of local queue metrics to users, providing detailed insights into workload 
processing specific to individual namespaces / tenants.

## Motivation

Metrics related to local queues are invaluable for batch users (usually with namespace-scoped permissions), as they provide 
essential visibility and historical trends about their workloads. Currently, metrics are available for ClusterQueues but 
not for namespace-scoped batch users. Cluster queue metrics are often ineffective for batch users and namespace admins 
because they are global and cannot be filtered by namespaces. Furthermore, accessing cluster-scoped metrics 
in secured Prometheus instances is generally restricted to cluster admin users with cluster level permissions across all namespaces and 
tenants. This restriction makes it challenging for batch users to obtain the specific metrics they need for effective workload 

> management and to gain insights into their workloads within their limited scope of access.

Reviewer: Could this KEP also specify how we will prevent users with insufficient permission accessing metrics they shouldn't be able to?

Reviewer 2: If this proposal is not suggesting how to prevent access from namespaces that shouldn't have the permission, then I would remove this phrase.

Reviewer: I agree, maybe we could narrow the scope of this KEP @varshaprasad96 ? At first glance managing permissions seems to be challenging, or it would require external mechanism

Author: Sorry for the late reply! I've been considering various options, and it's clear that publishing local queue metrics in each namespace could lead to complications. This approach would require multiple endpoints with respective Service Monitors or, if using a single central Service Monitor, the correct RBAC setup to allow client access. The centralised approach could still poses some issues, such as namespace admins potentially being able to view metrics from other namespaces.

One potential solution for cluster admins is to scaffold out a Service Monitor (SM) with metrics [labeling](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#metric_relabel_configs) for specific namespaces from the same service endpoint. This would enable a common service endpoint for all local queues. Admins could then provide a service account with right RBAC to specific batch user, restricting their access to that particular service monitor. 

However, this solution is difficult to implement in Kueue right away by figuring out the right set of scaffolds. This could probably be a customisation which can be documented for now. 

That being said, given the complexity, this topic may probably warrant a further brainstorming and deserves a separate KEP. For now, I'll remove this from the current discussion and update the document to indicate that for this iteration metrics will be exported in the controller namespace, alongside cluster queue metrics, at the same endpoint. Does that sound reasonable, or would we want to a separate endpoint for all local queue metrics?

Reviewer 2: Yes it does.

You can remove the per-namespace topic from the motivation and add a note in non-goals.

collection for local queues.

The `ControllerMetrics` that contain the option to configure metrics, will be extended as follows:

```go
type ControllerManager struct {
	...

	// Metrics contains the controller metrics configuration
	// +optional
	Metrics ControllerMetrics `json:"metrics,omitempty"`
	...
}

// ControllerMetrics defines the metrics configs.
type ControllerMetrics struct {
	...
	
	// EnableLocalQueueResources, if true the local queue resource usage and quotas
	// metrics will be reported.
	// +optional
	EnableLocalQueueResources bool `json:"enableLocalQueueResources,omitempty"`
	
	// LocalQueueMetricOptions specifies the configuration options for local queue metrics.
	LocalQueueMetricOptions *metricsOptions `json:"localQueueMetricOptions,omitempty"`
}

// metricsOptions defines the configuration options for local queue metrics.
// If left empty, then metrics will expose for all local queues across namespaces.
type metricsOptions struct {
	// NamespaceSelector can be used to select namespaces in which the local queues should
	// report metrics.
	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
	

> 	// QueueSelector can be used to choose the local queues that need metrics to be collected. 
> 	QueueSelector *metav1.LabelSelector `json:"queueSelector,omitempty"`

Reviewer 3: ```suggestion
	// LocalQueueSelector can be used to choose the local queues that need metrics to be collected. 
	LocalQueueSelector *metav1.LabelSelector `json:"localQueueSelector,omitempty"`
```

Let's naming this field as more accurately.

> 	QueueSelector *metav1.LabelSelector `json:"queueSelector,omitempty"`

Reviewer 3: What happens if the below situations?

1. only namespeceSelector is specified
2. only localQueueSelector is specified
3. both namespaceSelector and localQueueSelector are specified

Author: Had a predicate based filtering kind of approach in mind for this case. 

1. Only NS selector: All LocalQueues in the specific namespaces are selected. If it's nil, all the local queues across all namespaces are considered. 
2. LocalQueue selector: Local queues's matching the labels are selected across all namespaces. If not, all the local queues across namespaces are considered.
3. If both are specified: Both the selectors are applied to local queues (logical AND).

Had a Venn diagram in mind :)) Please let me know if I'm missing anything.

Author: Have added a configuration table for this knob in the KEP.

}
```

To reduce cardinality, and enable selection of metrics for local queues, the following
knobs will be available:

1. If `EnableLocalQueueResources` is false, then metrics will not be exposed.  
1. If `EnableLocalQueueResources` is true, and `LocalQueueMetricOptions` is **not** provided - metrics will be exposed for all local queues.
1. If `EnableLocalQueueResources` is true and `LocalQueueMetricOptions`  is provided - metrics will be collected for local queues that match specified label selectors.

### List of metrics for Local Queues:

In the first iteration, following are the list of metrics that would contain information on Local Queue statuses:

> | Metrics Name                             | Prometheus Type | Description                                                                                         | 

Reviewer 3: A part of [the clusterqueue metrics](https://kueue.sigs.k8s.io/docs/reference/metrics/#clusterqueue-status) is imported here like `local_queue_status`. Is there any reason that you want to drop a few metrics rather than clusterqueue ones?

Author: Updated the list to contain all the local_queue relevant ones. One additional metric is the `local_queue_wait_time_seconds` which is a sum of admission and reserved wait time.

	
	// LocalQueueSelector can be used to choose the local queues that need metrics to be collected. 
	LocalQueueSelector *metav1.LabelSelector `json:"localQueueSelector,omitempty"`
}
```

To reduce cardinality, and enable selection of metrics for local queues, the following
knobs will be available for `LocalQueueMetrics`:

| `Enable` | `NamespaceSelector` | `LocalQueueSelector` | Description                                                                                                        |
|----------|---------------------|----------------------|--------------------------------------------------------------------------------------------------------------------|
| False    | -                   | -                    | Metrics will not be exposed.                                                                                       |
| True     | -                   | -                    | Metrics for all local queues will be exposed.                                                                      |
| True     | Specified           | -                    | All LocalQueues in the specific namespaces that match the selector have metrics enabled.                           |

> | True     | -                   | Specified            | All LocalQueues matching the label selector have metrics enabled.                                                  | |

Reviewer 3: ```suggestion
| True     | -                   | Specified            | All LocalQueues matching the label selector have metrics enabled.                                                  |
```

👀

|------------------------------------------|-----------------|-----------------------------------------------------------------------------------------------------|
| local_queue_pending_workloads            | Gauge           | The number of pending workloads.                                                                    |
| local_queue_reserved_workloads_total     | Counter         | Total number of workloads in the LocalQueue reserving quota.                                        | 
| local_queue_admitted_workloads_total     | Counter         | Total number of admitted workloads.                                                                 |
| local_queue_resource_usage               | Gauge           | Total quantity of used quota per resource for a Local Queue.                                        |
| local_queue_evicted_workloads_total      | Counter         | The total number of evicted workloads in Local Queue.                                               |
| local_queue_reserved_wait_time_seconds   | Histogram       | The time between a workload was created or re-queued until it got quota reservation in local queue. |

> | local_admission_checks_wait_time_seconds | Histogram       | The time from when a workload got the quota reservation until admission in local queue.             |

Reviewer: ```suggestion
| local_queue_admission_checks_wait_time_seconds | Histogram       | The time from when a workload got the quota reservation until admission in local queue.             |
```

> | local_queue_admission_wait_time_seconds  | Histogram       | Time taken to accept the resources from cluster queue.                                              |

Reviewer: ```suggestion
| local_queue_admission_wait_time_seconds  | Histogram       | The time between a workload was created or requeued until admission.                                              |
```

| local_queue_status                       | Gauge           | Reports the status of the ClusterQueue.                                                             |

Each of these metrics will be augmented with relevant Prometheus labels, indicating the local queue name, namespace, 
and any other unique identifiers as required during implementation. They will be exported in the controller namespace, 
alongside cluster queue metrics, at the same endpoint.

### Test Plan

[X] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

#### Unit Tests

There are existing unit tests for prometheus metrics: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/metrics/metrics_test.go.
However, unit tests to ensure coverage for any additional local queue metrics will be added. 

> - `<package>`: `<date>` - `<test coverage>`

Reviewer 2: Please fill up this

}
```

To reduce cardinality, and enable selection of metrics for local queues, the following
knobs will be available for `LocalQueueMetrics`:

| `Enable` | `NamespaceSelector` | `LocalQueueSelector` | Description                                                                                                        |
|----------|---------------------|----------------------|--------------------------------------------------------------------------------------------------------------------|
| False    | -                   | -                    | Metrics will not be exposed.                                                                                       |
| True     | -                   | -                    | Metrics for all local queues will be exposed.                                                                      |
| True     | Specified           | -                    | All LocalQueues in the specific namespaces that match the selector have metrics enabled.                           |
| True     | -                   | Specified            | All LocalQueues matching the label selector have metrics enabled.                                                  |

> | True     | -                   | Specified            | All LocalQueues matching the label selector have metrics enabled.                                                  |

Reviewer: Duplicated

| local_queue_reserved_workloads_total     | Counter         | Total number of workloads in the LocalQueue reserving quota.                                        | 
| local_queue_admitted_workloads_total     | Counter         | Total number of admitted workloads.                                                                 |
| local_queue_resource_usage               | Gauge           | Total quantity of used quota per resource for a Local Queue.                                        |
| local_queue_evicted_workloads_total      | Counter         | The total number of evicted workloads in Local Queue.                                               |
| local_queue_admission_wait_time_seconds  | Histogram       | The time from when a workload got the quota reservation until admission in local queue.             |
| local_queue_reserved_wait_time_seconds   | Histogram       | The time between a workload was created or re-queued until it got quota reservation in local queue. |
| local_queue_wait_time_seconds            | Histogram       | Time taken to accept the resources from cluster queue.                                              |

> | local_admission_checks_wait_time_seconds | Histogram       | The time from when a workload got the quota reservation until admission.                            |

Reviewer: Could you please clarify the differences between `local_queue_admission_wait_time_seconds` and `local_admission_checks_wait_time_seconds`? Maybe `local_queue_admission_wait_time_seconds` could describe time from creating (or requeueing) a Workload until admission - similarly to [the clusterqueue metrics](https://kueue.sigs.k8s.io/docs/reference/metrics/#clusterqueue-status). In that case it would cover the `local_queue_wait_time_seconds` metric usage.

Author: @PBundyra As you mentioned and looking into the code, looks like we would just need these 3:

1. local_queue_reserved_wait_time_seconds: The time between when a workload was created or re-queued until it got a quota reservation in the local queue.
2. local_queue_admission_wait_time_seconds: The time from when a workload got the quota reservation until it was admitted in the local queue.
3. local_queue_wait_time_seconds: Time taken to accept the resources from the cluster queue, which is going to be local_queue_reserved_wait_time_seconds + local_queue_admission_wait_time_seconds.

This would cover workloads lifecycle of: getting created --> Got reservation of quota --> Admitted.

Reviewer: Thank you @varshaprasad96! Sounds good to me, the only change I suggest is naming. Let's stick with `local_admission_checks_wait_time_seconds` instead of `local_queue_admission_wait_time_seconds` as the second metric you mentioned; and `local_queue_admission_wait_time_seconds` instead of `local_queue_wait_time_seconds` as the third one.

That way it would be consistent with the clusterqueue metrics I linked above, and express more accurately what happens inside Kueue. So the final state for these metrics I suggest is:

1. `local_queue_reserved_wait_time_seconds`: [...]
2. `local_admission_checks_wait_time_seconds`: [...]
3. `local_queue_admission_wait_time_seconds`: [...]

Author: Sounds good! Have modified the naming as suggested.
